tests/integrations/raftstore/test_merge.rs

```diff
@@ -1385,3 +1387,140 @@ fn test_merge_pessimistic_locks_when_gap_is_too_large() {
+            .is_ok()
+    );
+    // Filter MsgTimeoutNow, so transfer leader will not actually succeed.
+    cluster.add_send_filter(CloneFilterFactory(
+        RegionPacketFilter::new(left.id, 2)
+            .msg_type(MessageType::MsgTimeoutNow)
+            .direction(Direction::Recv),
+    ));
+
+    let router = cluster.sim.rl().get_router(1).unwrap();
+    let from_peer = left.peers.iter().find(|p| p.store_id == 2).unwrap().clone();
+    let to_peer = left.peers.iter().find(|p| p.store_id == 1).unwrap().clone();
+    let mut msg = eraftpb::Message::default();
+    msg.set_msg_type(eraftpb::MessageType::MsgTransferLeader);
+    msg.set_log_term(snapshot.term.unwrap().into());
+    msg.set_from(from_peer.get_id());
+    msg.set_to(to_peer.get_id());
+    msg.set_context(vec![1].into());
+    let mut message = RaftMessage::default();
+    message.set_region_id(left.id);
+    message.set_from_peer(from_peer);
+    message.set_to_peer(to_peer);
+    message.set_region_epoch(left.get_region_epoch().clone());
+    message.set_message(msg);
+    router.send_raft_message(message).unwrap();
+
+    // Now, the region is in transfer leader state, it will fail to propose the locks
+    // or the PrepareMerge request.
+    cluster.merge_region(left.id, right.id, Callback::None);
+    thread::sleep(Duration::from_millis(150));
+
+    // But after that, the pessimistic locks status should remain unchanged.
+    assert_eq!(txn_ext.pessimistic_locks.read().status, LocksStatus::Normal);
+}
+
+#[test]
+fn test_merge_pessimistic_locks_repeated_merge() {
+    let mut cluster = new_server_cluster(0, 2);
+    configure_for_merge(&mut cluster);
+    cluster.cfg.pessimistic_txn.pipelined = true;
+    cluster.cfg.pessimistic_txn.in_memory = true;
+    let pd_client = Arc::clone(&cluster.pd_client);
+    pd_client.disable_default_operator();
+
+    cluster.run();
+
+    cluster.must_transfer_leader(1, new_peer(1, 1));
+
+    cluster.must_put(b"k1", b"v1");
+    cluster.must_put(b"k3", b"v3");
+
+    let region = cluster.get_region(b"k1");
+    cluster.must_split(&region, b"k2");
+    let left = cluster.get_region(b"k1");
+    let right = cluster.get_region(b"k3");
+
+    let snapshot = cluster.must_get_snapshot_of_region(left.id);
+    let txn_ext = snapshot.ext().get_txn_ext().unwrap().clone();
+    let lock = PessimisticLock {
+        primary: b"k1".to_vec().into_boxed_slice(),
+        start_ts: 10.into(),
+        ttl: 3000,
+        for_update_ts: 20.into(),
+        min_commit_ts: 30.into(),
+    };
+    assert!(
+        txn_ext
+            .pessimistic_locks
+            .write()
+            .insert(vec![(Key::from_raw(b"k1"), lock.clone())])
+            .is_ok()
+    );
+
+    // Filter MsgAppend, so the proposed PrepareMerge will not succeed
+    cluster.add_send_filter(CloneFilterFactory(
+        RegionPacketFilter::new(left.id, 2)
+            .msg_type(MessageType::MsgAppend)
+            .direction(Direction::Recv),
+    ));
```

Comment:
Propose is the procedure to generate `MsgAppend` it can still success even can't send the msg to other peers, how about using a fail point to return an error early when proposing to simulate the case.

Reply:
It doesn't need to really fail. Instead, I want it to succeed, but not until I remove the filters. 

This is to generate the case that a second PrepareMerge is proposed before the previous one succeeds. This is to test against this part:

```rust
            // If we just failed to propose PrepareMerge, the pessimistic locks status
            // may become MergingRegion incorrectly. So, we have to revert it here.
            // But we have to rule out the case when the region has successfully
            // proposed PrepareMerge or has been in merging, which is decided by
            // the boolean expression below.
            let is_merging = self.is_merging()
                || self
                    .cmd_epoch_checker
                    .last_cmd_index(AdminCmdType::PrepareMerge)
                    .is_some();
```

Reply:
I see, but filtering just one peer may not be enough as there are still majority of peers.

Reply:
I configure rhe simulated cluster to only have 2 nodes.